### DISCUSSION TITLE
Iterate though Dicom tag & change to uppercase

### DISCUSSION THREAD

**Discussion by Ruffdraft** [nextgenhealthcare/connect#4591] [Q&A] [answered]
Trying to modify each instance of ROIObservationLabel to uppercase using Javascript, but having issues as its not chaning the tags in the dicom file.

The following code is in the Source Transformer (Type Javascript) of a DICOM Listener;

```
var tmp = msg['tag30060085'].toString();
for each(msg in tmp)
{
  switch(msg)
   {
       //if it equals "lowercase", replace with "UPPERCASE"
     case "Prostate":
       msg['tag30060085'] = "PROSTATE";
       break;

     case "Bladder":
       msg['tag30060085'] = "BLADDER";
       break;
	}
}
```
tried this also, as its I'm unsure its able to read the sub tag30060085 of tag30060080

```
if (msg['tag30060080']['tag30060085'].includes("Prostate")) {
    tstr = "PROSTATE";
    msg['tag30060080']['tag30060085'] = tstr.toString();
    msg['tag30060080']['tag30060085']['@len'] = tstr.toString().length;
    msg['tag30060080']['tag30060085']['@tag'] = "30060085";
    msg['tag30060080']['tag30060085']['@vr'] = "SH";
}
```

thanks.

**Comment by tonygermano** [comment, marked as answer]
There's lots of stuff that won't work in there.
- `msg['tag30060085']` means to return a list of all child elements of `msg` named `tag30060085`. Taking the `toString()` value of that means you now have a string instead of an XMLList, so you can't iterate over it with the `for each`.
- Using `msg` as your loop variable is going to overwrite your actual `msg`.
- I'm not familiar with the DICOM tags, but if `tag30060085` is a sub-tag as you indicated in your second code block, it would not be found by `msg['tag30060085']`.
- This isn't wrong, but it's not necessary in your second example to call `tstr.toString()` because `tstr` is already a string, so you can just use it directly.

I took a stab at a solution based on the examples in your question. I have no idea if this will actually work.
```js
// find any tag30060085 in the document tree rather than only direct descendants
for each (var tag in msg..tag30060085) {
    // get string value of the tag
    var value = tag.toString();
    var uppercase = value.toUpperCase();
    if (value != uppercase) {
        // replace only the child text node instead of the entire tag to preserve attributes
        tag.setChildren(uppercase);
    }
}
```
It might help to put your channel storage setting in development mode, and then look at the transformed content of one of your messages in the message browser. This can help you see how the XML is actually being structured.

**Reply by Ruffdraft** [reply]
Thanks @tonygermano this worked first time. 

I need to modify the tag30060026 with the same info, so just incase this helps someone else.

See below this is working for me;

```
for each (var tag in msg..tag30060085, msg..tag30060026) {
    // get string value of the tag
    var value = tag.toString();
    var uppercase = value.toUpperCase();
    if (value != uppercase) {
        // replace only the child text node instead of the entire tag to preserve attributes
        tag.setChildren(uppercase);
    }
}
```
